core/src/main/java/org/geysermc/geyser/custom/items/GeyserCustomItemManager.java

```diff
@@ -0,0 +1,142 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @author GeyserMC
+ * @link https://github.com/GeyserMC/Geyser
+ */
+
+package org.geysermc.geyser.custom.items;
+
+import com.nukkitx.protocol.bedrock.data.inventory.ComponentItemData;
+import com.nukkitx.protocol.bedrock.packet.StartGamePacket;
+import it.unimi.dsi.fastutil.ints.Int2IntMap;
+import it.unimi.dsi.fastutil.ints.Int2IntOpenHashMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import org.geysermc.geyser.GeyserImpl;
+import org.geysermc.geyser.api.custom.items.CustomItemData;
+import org.geysermc.geyser.api.custom.items.CustomItemManager;
+import org.geysermc.geyser.api.custom.items.registration.CustomItemRegistrationType;
+import org.geysermc.geyser.api.custom.items.registration.CustomModelDataItemType;
+import org.geysermc.geyser.custom.GeyserCustomManager;
+import org.geysermc.geyser.registry.type.ItemMapping;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+public class GeyserCustomItemManager extends CustomItemManager {
+    private Map<String, List<GeyserCustomItemData>> customMappings = new HashMap<>();
+    private Int2ObjectMap<String> customIdMappings = new Int2ObjectOpenHashMap<>();
+
+    private int registeredCustomItems = 0;
+
+    @Override
+    public void registerCustomItem(String baseItem, CustomItemData customItemData) {
+        int nameExists = 0;
+        for (String name : this.customIdMappings.values()) {
+            String addName = GeyserCustomManager.CUSTOM_PREFIX + customItemData.name();
+            if (Pattern.matches("^" + addName +"(_([0-9])+)?$", name)) {
+                nameExists++;
+            }
+        }
+        if (nameExists != 0) {
+            GeyserImpl.getInstance().getLogger().warning("Custom item name '" + customItemData.name() + "' already exists and was registered again!");
+        }
+
+        GeyserCustomItemData registeredItem = CustomItemsRegistryPopulator.addToRegistry(baseItem, customItemData, nameExists, this);
+        if (registeredItem != null && registeredItem.mappings().size() > 0) {
+            if (this.customMappings.containsKey(baseItem)) {
```

Comment:
You can replace this whole if statement with `this.customMappings.computeIfAbsent(baseItem, ArrayList::new).add(registeredItem);`